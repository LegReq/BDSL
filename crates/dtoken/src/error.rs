//! Errors returned when generating code.

use std::{
    fmt::{self, Display},
    num::{ParseFloatError, ParseIntError},
};

use tinyjson::JsonParseError;

/// Error type returned when the code generation failed for some reason.
#[derive(Debug)]
pub enum BuildError {
    Parse(Error),
    Fmt(std::io::Error),
    Read(std::io::Error),
    Write(std::io::Error),
    Var(std::env::VarError),
    JsonParse(tinyjson::JsonParseError),
    #[cfg(feature = "toml")]
    TomlParse(toml_span::Error),
    #[cfg(feature = "ason")]
    AsonParse(ason::AsonError),
    #[cfg(feature = "jsonc")]
    JsoncParse(jsonc_parser::errors::ParseError),
}

impl std::error::Error for BuildError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            BuildError::Parse(v) => Some(v),
            BuildError::Var(v) => Some(v),
            BuildError::JsonParse(v) => Some(v),
            #[cfg(feature = "toml")]
            BuildError::TomlParse(v) => Some(v),
            #[cfg(feature = "ason")]
            BuildError::AsonParse(v) => Some(v),
            #[cfg(feature = "jsonc")]
            BuildError::JsoncParse(v) => Some(v),
            BuildError::Fmt(v) | BuildError::Read(v) | BuildError::Write(v) => Some(v),
        }
    }
}

impl Display for BuildError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            BuildError::Parse(error) => write!(f, "failed to parse file: {error}"),
            BuildError::Fmt(error) => write!(f, "failed to run rustfmt: {error}"),
            BuildError::Read(error) => write!(f, "failed to read file: {error}"),
            BuildError::Write(error) => write!(f, "failed to write file: {error}"),
            BuildError::Var(error) => write!(f, "failed to read environment variable: {error}"),
            BuildError::JsonParse(error) => write!(f, "failed to parse json file: {error}"),
            #[cfg(feature = "toml")]
            BuildError::TomlParse(error) => write!(f, "failed to parse toml file: {error}"),
            #[cfg(feature = "ason")]
            BuildError::AsonParse(error) => write!(f, "failed to parse ason file: {error}"),
            #[cfg(feature = "jsonc")]
            BuildError::JsoncParse(error) => write!(f, "failed to parse jsonc file: {error}"),
        }
    }
}

impl From<Error> for BuildError {
    fn from(value: Error) -> Self {
        Self::Parse(value)
    }
}

impl From<JsonParseError> for BuildError {
    fn from(source: JsonParseError) -> Self {
        Self::JsonParse(source)
    }
}

#[cfg(feature = "toml")]
impl From<toml_span::Error> for BuildError {
    fn from(source: toml_span::Error) -> Self {
        Self::TomlParse(source)
    }
}

#[cfg(feature = "ason")]
impl From<ason::AsonError> for BuildError {
    fn from(source: ason::AsonError) -> Self {
        Self::AsonParse(source)
    }
}

#[cfg(feature = "jsonc")]
impl From<jsonc_parser::errors::ParseError> for BuildError {
    fn from(source: jsonc_parser::errors::ParseError) -> Self {
        Self::JsoncParse(source)
    }
}

impl From<std::env::VarError> for BuildError {
    fn from(error: std::env::VarError) -> Self {
        Self::Var(error)
    }
}

/// Error type returned when a parsing error occurs.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Error {
    Property(&'static str, Box<Error>),
    Kind(String, Box<Error>),

    MustExist,

    ExpectedString,
    ExpectedNumber,
    ExpectedArray,
    ExpectedObject,
    UnexpectedType,

    CollectionEmpty,
    CollectionLength(usize),
    ExpectedItemString,
    ExpectedItemNumber,
    ExpectedItemObject,

    NumberMustBePositive,
    NumberWithin(isize, isize),

    InvalidNumber(String),
    InvalidUnit(&'static [&'static str]),
    InvalidFormat(&'static str),
    MissingToken(char),
}

impl Error {
    #[must_use]
    pub fn prop(property: &'static str, error: Self) -> Self {
        Self::Property(property, Box::new(error))
    }

    #[must_use]
    pub fn kind(kind: String, err: Self) -> Self {
        Self::Kind(kind, Box::new(err))
    }
}

impl From<ParseFloatError> for Error {
    fn from(err: ParseFloatError) -> Self {
        Self::InvalidNumber(err.to_string())
    }
}

impl From<ParseIntError> for Error {
    fn from(err: ParseIntError) -> Self {
        Self::InvalidNumber(err.to_string())
    }
}

impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Error::Property(_, source) => Some(source.as_ref()),
            _ => None,
        }
    }
}

impl Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Property(prop, err) => write!(f, "property '{prop}' error: {err}"),
            Self::Kind(kind, err) => write!(f, "value error for $type '{kind}': {err}"),
            Self::MustExist => write!(f, "must exist"),
            Self::ExpectedString
            | Self::ExpectedNumber
            | Self::ExpectedArray
            | Self::ExpectedObject => {
                let ty = match self {
                    Self::ExpectedString => "string",
                    Self::ExpectedNumber => "number",
                    Self::ExpectedArray => "array",
                    _ => unreachable!(),
                };

                write!(f, "must be of type {ty}")
            }
            Self::ExpectedItemString | Self::ExpectedItemNumber | Self::ExpectedItemObject => {
                let ty = match self {
                    Self::ExpectedItemString => "string",
                    Self::ExpectedItemNumber => "number",
                    Self::ExpectedItemObject => "object",
                    _ => unreachable!(),
                };

                write!(f, "collection items must be of type {ty}")
            }
            Self::MissingToken(char) => write!(f, "missing token: {char}"),
            Self::NumberWithin(start, end) => write!(f, "number must be >= {start} <= {end}"),
            Self::UnexpectedType => write!(f, "unexpected type"),
            Self::NumberMustBePositive => write!(f, "must be a positive number"),
            Self::InvalidNumber(err) => write!(f, "invalid number format: {err}"),
            Self::InvalidUnit(units) => {
                write!(f, "invalid unit, must be one of {}", units.join(", "))
            }
            Self::InvalidFormat(str) => write!(f, "invalid format: {str}"),
            Self::CollectionEmpty => write!(f, "collection must not be empty"),
            Self::CollectionLength(len) => write!(f, "collection must contain {len} elements"),
        }
    }
}
