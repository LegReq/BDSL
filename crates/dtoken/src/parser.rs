//! Translations files parsing
//!
//! Files are parsed as [TranslationData] from a provided [JsonValue].
//! Parsed keys are represented as [TranslationKey].

pub mod group;
pub mod token;
pub mod types;

// use std::collections::{HashMap, HashSet};
//
// use lazy_static::lazy_static;
// use regex::Regex;
// use tinyjson::JsonValue;
//
// use crate::{builder::TokensId, error::ParseError};
//
// /// Data structure containing all translation keys
// ///
// /// This struct should be initialized with the fallback language,
// /// then keys will be populated with other languages using the [`parse_file`] method.
// ///
// /// [`parse_file`]: Self::parse_file
// #[derive(Debug, Clone, PartialEq, Eq, Default)]
// pub(crate) struct TokensData {
//     /// Parsed translation keys
//     pub(crate) keys: HashMap<String, TranslationKey>,
// }
//
// impl TokensData {
//     /// Parse a language file and insert its content into the current [`TranslationData`]
//     pub(crate) fn parse_file(
//         &mut self,
//         group: TokensId,
//         file: JsonValue,
//     ) -> Result<(), ParseError> {
//         let parsed = ParsedFile::parse(file)?;
//
//         for (key, parsed) in parsed.keys {
//             match self.keys.get_mut(&key) {
//                 Some(translation_key) => {
//                     let data = ParsedKeyData {
//                         language: group.clone(),
//                         key: &key,
//                         parsed,
//                     };
//                     translation_key.insert_parsed(data)?
//                 }
//                 None => println!(
//                     "cargo:warning=Key `{}` exists in {} but not in fallback language",
//                     key, group
//                 ),
//             };
//         }
//
//         Ok(())
//     }
// }
//
// /// A parsed group key
// ///
// /// This enum can be constructed by parsing a translation file with [TranslationData].
// #[derive(Debug, Clone, PartialEq, Eq)]
// pub enum TranslationKey {
//     Simple(SimpleKey),
//     Formatted(FormattedKey),
// }
//
// impl TranslationKey {
//     // /// Initialize a new [TranslationKey] from a [`ParsedKey`]
//     // pub(crate) fn from_parsed(parsed: ParsedKey) -> Self {
//     //     match parsed {
//     //         ParsedKey::Simple(value) => TranslationKey::Simple(SimpleKey {
//     //             fallback: value,
//     //             others: HashMap::new(),
//     //         }),
//     //         ParsedKey::Formatted { value, parameters } => TranslationKey::Formatted(FormattedKey {
//     //             fallback: value,
//     //             others: HashMap::new(),
//     //             parameters,
//     //         }),
//     //     }
//     // }
//
//     /// Inserts a new raw [`ParsedKey`] in this [`TranslationKey`]
//     fn insert_parsed(&mut self, data: ParsedKeyData) -> Result<(), ParseError> {
//         match self {
//             TranslationKey::Simple(inner) => inner.insert_parsed(data),
//             TranslationKey::Formatted(inner) => inner.insert_parsed(data),
//         }
//     }
// }
//
// #[derive(Debug, Clone, PartialEq, Eq)]
// /// Simple string key, without any formatting or plurals
// pub struct SimpleKey {
//     /// The key value for the fallback language
//     pub(crate) fallback: String,
//     /// Key values for other languages
//     pub(crate) others: HashMap<TokensId, String>,
// }
//
// impl SimpleKey {
//     /// Inserts a new raw [`ParsedKey`] in this [`SimpleKey`]
//     fn insert_parsed(&mut self, data: ParsedKeyData) -> Result<(), ParseError> {
//         match data.parsed {
//             ParsedKey::Simple(value) => self.others.insert(data.language, value),
//             _ => {
//                 return Err(ParseError::InvalidType {
//                     key: data.key.into(),
//                     expected: "string",
//                 })
//             }
//         };
//
//         Ok(())
//     }
// }
//
// #[derive(Debug, Clone, PartialEq, Eq)]
// /// Simple string key with formatting
// pub struct FormattedKey {
//     /// The key value for the fallback language
//     pub(crate) fallback: String,
//     /// Key values for other languages
//     pub(crate) others: HashMap<TokensId, String>,
//     /// List of parameters in the value
//     pub(crate) parameters: HashSet<String>,
// }
//
// impl FormattedKey {
//     /// Inserts a new [`ParsedKey`] in this [`SimpleKey`]
//     fn insert_parsed(&mut self, data: ParsedKeyData) -> Result<(), ParseError> {
//         let (value, parameters) = match data.parsed {
//             ParsedKey::Formatted { value, parameters } => (value, parameters),
//             _ => {
//                 return Err(ParseError::InvalidType {
//                     key: data.key.into(),
//                     expected: "formatted string",
//                 })
//             }
//         };
//
//         if parameters == self.parameters {
//             self.others.insert(data.language, value);
//             Ok(())
//         } else {
//             let missing: Vec<_> = self.parameters.difference(&parameters).cloned().collect();
//             let unknown: Vec<_> = parameters.difference(&self.parameters).cloned().collect();
//
//             Err(ParseError::InvalidParameters {
//                 key: data.key.into(),
//                 missing,
//                 unknown,
//             })
//         }
//     }
// }
//
// /// Raw representation of a parsed file
// #[derive(Debug, Clone, PartialEq, Eq)]
// struct ParsedFile {
//     keys: HashMap<String, ParsedKey>,
// }
//
// impl ParsedFile {
//     /// Parse a JSON [`JsonValue`] as a translations file
//     fn parse(file: JsonValue) -> Result<Self, ParseError> {
//         let input = match file {
//             JsonValue::Object(map) => map,
//             _ => return Err(ParseError::InvalidRoot),
//         };
//
//         let mut keys = HashMap::with_capacity(input.len());
//         for (key, value) in input {
//             let parsed = ParsedKey::parse(&key, value)?;
//             keys.insert(key, parsed);
//         }
//
//         Ok(ParsedFile { keys })
//     }
// }
//
// /// Raw representation of a parsed key
// #[derive(Debug, Clone, PartialEq, Eq)]
// enum ParsedKey {
//     /// Simple string key
//     Simple(String),
//     /// String key with formatted values
//     ///
//     /// Example : `Hello {name}!`
//     Formatted {
//         /// The raw key value
//         value: String,
//         /// List of parameters in the value
//         parameters: HashSet<String>,
//     },
// }
//
// impl ParsedKey {
//     /// Parse a JSON [`Value`] as a key
//     fn parse(key: &str, value: JsonValue) -> Result<Self, ParseError> {
//         match value {
//             JsonValue::String(value) => Ok(Self::parse_string(value)),
//             _ => Err(ParseError::InvalidValue { key: key.into() }),
//         }
//     }
//
//     fn parse_string(value: String) -> Self {
//         lazy_static! {
//             static ref RE: Regex = Regex::new(r"\{([a-z_]+)\}").unwrap();
//         }
//
//         let matches: HashSet<_> = RE
//             .captures_iter(&value)
//             .map(|capture| capture[1].to_string())
//             .collect();
//
//         if matches.is_empty() {
//             Self::Simple(value)
//         } else {
//             Self::Formatted {
//                 value,
//                 parameters: matches,
//             }
//         }
//     }
// }
//
// /// Data associated with a parsed key.
// ///
// /// Used in [`TranslationKey::insert_parsed`].
// #[derive(Debug, Clone, PartialEq, Eq)]
// struct ParsedKeyData<'a> {
//     language: TokensId,
//     key: &'a str,
//     parsed: ParsedKey,
// }
