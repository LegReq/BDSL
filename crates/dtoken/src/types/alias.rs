//! Instead of having explicit values, tokens can reference the value of another token. To put it
//! another way, a token can be an alias for another token. This spec considers the terms "alias"
//! and "reference" to be synonyms and uses them interchangeably.
//!
//! Aliases are useful for:
//!
//! - Expressing design choices
//! - Eliminating repetition of values in token files (DRYing up the code)
//!
//! For a design token to reference another, its value MUST be a string containing the
//! period-separated (.) path to the token it's referencing enclosed in curly brackets.
//!
//! For example:
//! Example 15
//!
//! ```json,ignore
//! {
//!   "group name": {
//!     "token name": {
//!       "$value": 1234,
//!       "$type": "number"
//!     }
//!   },
//!   "alias name": {
//!     "$value": "{group name.token name}"
//!   }
//! }
//! ```
//!
//! When a tool needs the actual value of a token it MUST resolve the reference - i.e. lookup the
//! token being referenced and fetch its value. In the above example, the "alias name" token's
//! value would resolve to 1234 because it references the token whose path is {group name.token
//! name} which has the value 1234.
//!
//! Tools SHOULD preserve references and therefore only resolve them whenever the actual value
//! needs to be retrieved. For instance, in a design tool, changes to the value of a token being
//! referenced by aliases SHOULD be reflected wherever those aliases are being used.
//!
//! Aliases MAY reference other aliases. In this case, tools MUST follow each reference until they
//! find a token with an explicit value. Circular references are not allowed. If a design token
//! file contains circular references, then the value of all tokens in that chain is unknown and an
//! appropriate error or warning message SHOULD be displayed to the user.
#[derive(Debug, Clone, PartialEq)]
pub struct Alias {
    pub path_segments: Vec<String>,
}

impl Alias {
    pub fn from_str(s: &str) -> Option<Self> {
        if s.starts_with('{') && s.ends_with('}') {
            let path = &s[1..s.len() - 1]; // Remove the curly braces
            let path_segments = path.split('.').map(String::from).collect();
            Some(Self { path_segments })
        } else {
            None
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_alias_from_str() {
        #[rustfmt::skip]
        let test_cases = vec![
            ("{foo.bar}", Some(Alias { path_segments: vec!["foo".to_string(), "bar".to_string()] })),
            ("{abc.xyz}", Some(Alias { path_segments: vec!["abc".to_string(), "xyz".to_string()] })),
            ("{token}", Some(Alias { path_segments: vec!["token".to_string()] })),
            ("not_an_alias", None),
            ("{}valid{}", Some(Alias { path_segments: vec!["}valid{".to_string()] })),
            ("{foo.bar", None),
            ("foo.bar}", None),
        ];

        for (input, expected) in test_cases {
            let result = Alias::from_str(input);
            assert_eq!(result, expected);
        }
    }
}
