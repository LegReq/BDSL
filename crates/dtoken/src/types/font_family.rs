//! Represents a font name or an array of font names (ordered from most to least preferred). The
//! $type property MUST be set to the string fontFamily. The value MUST either be a string value
//! containing a single font name or an array of strings, each being a single font name.
//!
//! For example:
//!
//! EXAMPLE 19
//! ```json,ignore
//! {
//!   "Primary font": {
//!     "$value": "Comic Sans MS",
//!     "$type": "fontFamily"
//!   },
//!   "Body font": {
//!     "$value": ["Helvetica", "Arial", "sans-serif"],
//!     "$type": "fontFamily"
//!   }
//! }
//! ```
//!
//! See: <https://tr.designtokens.org/format/#font-family>.

use tinyjson::JsonValue;

/// See module docs.
#[derive(Debug, Clone, PartialEq)]
pub struct FontFamily {
    pub primary: String,
    pub fallbacks: Vec<String>,
}
impl FontFamily {
    pub fn from_str(s: &str) -> Self {
        Self {
            primary: s.to_owned(),
            fallbacks: vec![],
        }
    }
    pub fn from_slice(value: &[JsonValue]) -> Option<Self> {
        value
            .into_iter()
            .map(|val| val.get::<String>().map(ToOwned::to_owned))
            .collect::<Option<Vec<_>>>()?
            .split_first()
            .map(|(primary, fallbacks)| FontFamily {
                primary: primary.to_owned(),
                fallbacks: fallbacks.to_vec(),
            })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_from_str() {
        let test_cases = vec![
            (
                "Arial",
                FontFamily {
                    primary: "Arial".to_owned(),
                    fallbacks: vec![],
                },
            ),
            (
                "Helvetica, Arial, sans-serif",
                FontFamily {
                    primary: "Helvetica, Arial, sans-serif".to_owned(),
                    fallbacks: vec![],
                },
            ),
            (
                "Roboto, 'Noto Sans', sans-serif",
                FontFamily {
                    primary: "Roboto, 'Noto Sans', sans-serif".to_owned(),
                    fallbacks: vec![],
                },
            ),
        ];

        for (input, expected) in test_cases {
            let result = FontFamily::from_str(input);
            assert_eq!(result, expected);
        }
    }

    #[test]
    fn test_from_slice() {
        use JsonValue::{Number, String};

        let test_cases = vec![
            (
                vec![String("Arial".to_owned())],
                Some(FontFamily {
                    primary: "Arial".to_owned(),
                    fallbacks: vec![],
                }),
            ),
            (
                vec![
                    String("Helvetica".to_owned()),
                    String("Arial".to_owned()),
                    String("sans-serif".to_owned()),
                ],
                Some(FontFamily {
                    primary: "Helvetica".to_owned(),
                    fallbacks: vec!["Arial".to_owned(), "sans-serif".to_owned()],
                }),
            ),
            (
                vec![
                    String("Roboto".to_owned()),
                    String("'Noto Sans'".to_owned()),
                    String("sans-serif".to_owned()),
                ],
                Some(FontFamily {
                    primary: "Roboto".to_owned(),
                    fallbacks: vec!["'Noto Sans'".to_owned(), "sans-serif".to_owned()],
                }),
            ),
            (
                vec![Number(12.)],
                None, // Invalid JSON value
            ),
        ];

        for (input, expected) in test_cases {
            let result = FontFamily::from_slice(&input);
            assert_eq!(result, expected);
        }
    }
}
