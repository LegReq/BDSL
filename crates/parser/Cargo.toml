[package]
name = "bnb-parser"
version = "0.1.0"
edition = "2021"

[dependencies]
thiserror = { version = "1", default-features = false }

[dev-dependencies]
indoc = { version = "2", default-features = false }
insta = { version = "1", default-features = false, features = ["colors"] }
