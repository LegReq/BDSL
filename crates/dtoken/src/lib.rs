//! Code generation for the Rosetta i18n library.
//!
//! # Usage
//!
//! Code generation works within [build script]. You only need to configure source files. Please
//! read the [Design Tokens specification] for more information.
//!
//! ```no_run
//! dtoken::build("design_tokens.json").unwrap()
//! ```
//!
//! [build script]: https://doc.rust-lang.org/cargo/reference/build-scripts.html
//! [Design Tokens specification]: https://tr.designtokens.org

pub mod error;
pub mod types;

pub mod parser;

#[cfg(feature = "build")]
use std::{collections::HashMap, path::Path};

#[cfg(feature = "build")]
use convert_case::{Case, Casing};
#[cfg(feature = "build")]
use error::BuildError;
#[cfg(feature = "build")]
use parser::{group::Group, types::DesignTokens};
#[cfg(feature = "build")]
use proc_macro2::{Ident, Span, TokenStream};
#[cfg(feature = "build")]
use quote::quote;
#[cfg(feature = "build")]
use types::alias::Alias;
use types::font_family::FontFamily;

#[cfg(feature = "build")]
use crate::{
    parser::{token::Value, types::TokenOrGroup},
    types::{color::Color, number::Number},
};

/// Validate configuration and build a [`TokensConfig`]
#[cfg(feature = "build")]
pub fn build(path: impl AsRef<str>) -> Result<(), BuildError> {
    use tinyjson::JsonValue;

    let content = std::fs::read_to_string(path.as_ref())?;
    let json: JsonValue = content.parse()?;
    let map = json.get().ok_or(BuildError::Parse)?;

    let tokens = DesignTokens::from_map(map).ok_or(BuildError::Parse)?;
    let code = generate(tokens);

    let output = Path::new(&std::env::var("OUT_DIR")?).join("design_tokens.rs");

    std::fs::write(&output, code.to_string())?;

    #[cfg(all(feature = "rustfmt"))]
    rustfmt(&output)?;

    Ok(())
}

#[cfg(feature = "build")]
fn generate(tokens: DesignTokens) -> TokenStream {
    Generator::new(tokens).generate()
}

#[cfg(feature = "build")]
struct Generator {
    root: Group,
}

#[cfg(feature = "build")]
impl Generator {
    fn new(tokens: DesignTokens) -> Self {
        let root = Group {
            items: tokens.items.clone(),
            description: Some(" Root-level Design Tokens type".to_owned()),
            default_type: None,
            extensions: HashMap::new(),
        };

        Self { root }
    }

    fn generate(&self) -> TokenStream {
        let gen = self.module_impl("DesignTokens", &self.root);
        let instance = self.group_instance("DesignTokens", &self.root, vec![]);

        quote! {
            pub fn design_tokens() -> design_tokens::DesignTokens {
                #instance
            }

            #gen
        }
    }

    fn group_instance(&self, item: &str, group: &Group, mut parents: Vec<Ident>) -> TokenStream {
        let module_name = Ident::new(&item.to_case(Case::Snake), Span::call_site());
        let group_name = Ident::new(&item.to_case(Case::Pascal), Span::call_site());
        parents.push(module_name.clone());

        let mut items: Vec<_> = group.items.iter().collect();
        items.sort_by_key(|(k, _)| k.to_owned());

        let mut fields = vec![];
        let mut values = vec![];
        for (name, token_or_group) in &items {
            let (field, value) = self.field_instance(name, token_or_group, parents.clone());

            fields.push(field);
            values.push(value);
        }

        quote! {
            #(#parents::)* #group_name {
                #( #fields: #values,)*
            }
        }
        // quote! {
        //     DesignTokens {
        //         alias_name: dtokens::types::number::Number(123.456),
        //         group_name: group_name::GroupName {
        //             token_name: dtokens::types::number::Number(123,456),
        //         },
        //     }
        // }
    }

    fn field_instance(
        &self,
        field: &String,
        kind: &TokenOrGroup,
        parents: Vec<Ident>,
    ) -> (Ident, TokenStream) {
        let key = Ident::new(&field.to_case(Case::Snake), Span::call_site());
        let value = match kind {
            TokenOrGroup::Token(token) => self.token_value(&token.value),
            TokenOrGroup::Group(group) => self.group_instance(field, group, parents),
        };

        (key, value)
    }

    fn module_impl(&self, item: &str, group: &Group) -> TokenStream {
        let module = Ident::new(&item.to_case(Case::Snake), Span::call_site());
        let gen = self.group_impl(item, group);

        quote! {
            pub mod #module {
                #gen
            }
        }
    }

    fn group_impl(&self, item: &str, group: &Group) -> TokenStream {
        let group_name = Ident::new(&item.to_case(Case::Pascal), Span::call_site());
        let description = group.description.clone().unwrap_or_default();

        let mut items: Vec<_> = group.items.iter().collect();
        items.sort_by_key(|(k, _)| k.to_owned());

        let mut nested = vec![];
        for (name, group_item) in &items {
            let group = self.token_or_group_impl(name, group_item);
            nested.push(group);
        }

        let mut fields = vec![];
        let mut types = vec![];
        let mut descs = vec![];
        for (name, token_or_group) in &items {
            let (field, kind) = self.struct_field(name, token_or_group);
            let desc = token_or_group.description().unwrap_or_default();

            fields.push(field);
            types.push(kind);
            descs.push(desc);
        }

        quote! {
            #[doc = #description]
            #[derive(Debug)]
            pub struct #group_name {
                #( #[doc = #descs] pub #fields: #types,)*
            }

            #(#nested)*
        }
    }

    fn token_or_group_impl(&self, item: &str, token_or_group: &TokenOrGroup) -> TokenStream {
        match token_or_group {
            TokenOrGroup::Token(_) => return quote! {},
            TokenOrGroup::Group(group) => self.module_impl(item, group),
        }
    }

    fn struct_field(&self, field: &String, kind: &TokenOrGroup) -> (Ident, TokenStream) {
        let key = Ident::new(&field.to_case(Case::Snake), Span::call_site());
        let value = match kind {
            TokenOrGroup::Token(token) => self.token_kind(&token.value),
            TokenOrGroup::Group(_) => {
                let module = Ident::new(&field.to_case(Case::Snake), Span::call_site());
                let tail = Ident::new(&field.to_case(Case::Pascal), Span::call_site());
                quote! { #module :: #tail }
            }
        };

        (key, value)
    }

    fn alias_type(&self, alias: &Alias) -> Result<TokenStream, String> {
        let mut reference = &TokenOrGroup::Group(self.root.clone());

        for key in &alias.path_segments {
            reference = match reference {
                TokenOrGroup::Token(_) => {
                    return Err(format!(
                        "alias path segment {} points to value, but group was expected.",
                        key
                    ));
                }
                TokenOrGroup::Group(group) => match group.items.get(key) {
                    Some(token_or_group) => token_or_group,
                    None => {
                        return Err(format!(
                            "alias target missing: {{{}}}",
                            alias.path_segments.join(".")
                        ));
                    }
                },
            };
        }

        match reference {
            TokenOrGroup::Token(token) => Ok(self.token_kind(&token.value)),
            TokenOrGroup::Group(_) => Err(format!(
                "alias {{{}}} must point to a value, but instead points to a group.",
                alias.path_segments.join(".")
            )),
        }
    }

    fn token_kind(&self, value: &Value) -> TokenStream {
        let kind = match value {
            Value::Color(_) => "Color",
            Value::Dimension(_) => "Dimension",
            Value::FontFamily(_) => "FontFamily",
            Value::FontWeight(_) => "FontWeight",
            Value::Duration(_) => "Duration",
            Value::CubicBezier(_) => "CubicBezier",
            Value::Number(_) => "Number",
            Value::StrokeStyle(_) => "StrokeStyle",
            Value::Border(_) => "Border",
            Value::Transition(_) => "Transition",
            Value::Shadow(_) => "Shadow",
            Value::Gradient(_) => "Gradient",
            Value::Typography(_) => "Typography",
            Value::Alias(alias) => return self.alias_type(alias).unwrap(),
        };

        // let crate_name = Ident::new(&std::env::var("CARGO_PKG_NAME").unwrap(), Span::call_site());
        let module = Ident::new(&kind.to_case(Case::Snake), Span::call_site());
        let kind = Ident::new(kind, Span::call_site());

        quote! {
            dtoken::types::#module::#kind
        }
    }

    fn alias_value(&self, alias: &Alias) -> Result<TokenStream, String> {
        let mut reference = &TokenOrGroup::Group(self.root.clone());

        for key in &alias.path_segments {
            reference = match reference {
                TokenOrGroup::Token(_) => {
                    return Err(format!(
                        "alias path segment {} points to value, but group was expected.",
                        key
                    ));
                }
                TokenOrGroup::Group(group) => match group.items.get(key) {
                    Some(token_or_group) => token_or_group,
                    None => {
                        return Err(format!(
                            "alias target missing: {{{}}}",
                            alias.path_segments.join(".")
                        ));
                    }
                },
            };
        }

        match reference {
            TokenOrGroup::Token(token) => Ok(self.token_value(&token.value)),
            TokenOrGroup::Group(_) => Err(format!(
                "alias {{{}}} must point to a value, but instead points to a group.",
                alias.path_segments.join(".")
            )),
        }
    }

    fn token_value(&self, value: &Value) -> TokenStream {
        let kind = self.token_kind(value);
        let value = match value {
            Value::Color(Color { r, g, b, a }) => quote! { { r: #r, g: #g, b: #b, a: #a } },
            Value::Number(Number(v)) => quote! { (#v) },
            Value::Alias(alias) => return self.alias_value(alias).unwrap(),
            Value::FontFamily(FontFamily { primary, fallbacks }) => {
                quote! { {
                    primary: #primary.to_owned(),
                    fallbacks: vec![#( #fallbacks.to_owned(),)*],
                } }
            }
            v => todo!("{:?}", v),
            // Value::Dimension(v) => quote! { #v },
            // Value::FontWeight(v) => quote! { #v },
            // Value::Duration(v) => quote! { #v },
            // Value::CubicBezier(v) => quote! { #v },
            // Value::StrokeStyle(v) => quote! { #v },
            // Value::Border(v) => quote! { #v },
            // Value::Transition(v) => quote! { #v },
            // Value::Shadow(v) => quote! { #v },
            // Value::Gradient(v) => quote! { #v },
            // Value::Typography(v) => quote! { #v },
        };

        quote! {
            #kind #value
        }
    }
}

// fn generate_token_impl(item: &str, token: &Token) -> TokenStream {
//     let item = Ident::new(&item.to_case(Case::Pascal), Span::call_site());
//
//     quote! {
//         pub struct #item {
//
//         }
//     }
// }

// fn generate_token_or_group(name: &str, token_or_group: &TokenOrGroup) -> TokenStream {
//     struct Data {
//         field: String,
//         kind: String,
//         value: Value,
//     }
//
//     let name = name.to_case(Case::Pascal);
//     let mut fields = vec![];
//     let mut types = vec![];
//     // let mut values = vec![];
//
//     match token_or_group {
//         TokenOrGroup::Token(token) => {
//             if let Some(description) = &token.description {
//                 fields.push("description".to_owned());
//                 types.push("String".to_owned());
//                 // values.push(description.to_owned());
//             }
//         }
//         TokenOrGroup::Group(group) => {}
//     };
//
//     fields.sort();
//     types.sort();
//
//     dbg!(&fields);
//     dbg!(&types);
//
//     quote! {
//         pub struct #name;
//     }
// }

/// Format a file with rustfmt
#[cfg(all(feature = "rustfmt", feature = "build"))]
fn rustfmt(path: &Path) -> Result<(), BuildError> {
    use std::process::Command;

    Command::new(std::env::var("RUSTFMT").unwrap_or_else(|_| "rustfmt".to_string()))
        .args(&["--emit", "files"])
        .arg(path)
        .output()
        .map_err(BuildError::Fmt)?;

    Ok(())
}

#[cfg(all(test, feature = "build"))]
mod tests {
    use indoc::indoc;
    use syn::File;
    use tinyjson::JsonValue;

    use super::*;

    #[test]
    fn test_examples() {
        let test_cases = vec![indoc! {r#"
                {
                  "group name": {
                    "token name": {
                      "$value": 1234,
                      "$type": "number"
                    }
                  },
                  "alias name": {
                    "$value": "{group name.token name}"
                  }
                }
            "#}];

        for case in test_cases {
            let value: JsonValue = case.parse().unwrap();
            let tokens = DesignTokens::from_map(value.get().unwrap()).unwrap();

            let tokens = generate(tokens);
            let abstract_file: File =
                syn::parse2(tokens.clone()).unwrap_or_else(|err| panic!("{err}:\n\n{}", tokens));
            let code = prettyplease::unparse(&abstract_file);

            insta::assert_snapshot!(code.to_string());
        }
    }
}
