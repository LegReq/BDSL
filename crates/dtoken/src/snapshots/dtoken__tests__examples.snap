---
source: crates/dtoken/src/lib.rs
expression: code.to_string()
---
pub fn design_tokens() -> design_tokens::DesignTokens {
    design_tokens::DesignTokens {
        alias_name: dtoken::types::number::Number(1234f64),
        group_name: design_tokens::group_name::GroupName {
            token_name: dtoken::types::number::Number(1234f64),
        },
    }
}
pub mod design_tokens {
    /// Root-level Design Tokens type
    #[derive(Debug)]
    pub struct DesignTokens {
        ///
        pub alias_name: dtoken::types::number::Number,
        ///
        pub group_name: group_name::GroupName,
    }
    pub mod group_name {
        ///
        #[derive(Debug)]
        pub struct GroupName {
            ///
            pub token_name: dtoken::types::number::Number,
        }
    }
}

